idaes/surrogate/pysmo_surrogate.py

```diff
@@ -0,0 +1,828 @@
+
+    def save(self, stream: io.TextIOBase):
+        """Save this surrogate to the provided output stream so the model can be used later.
+
+        Args:
+           stream: Output stream for serialized surrogate object.
+
+        Returns:
+            None
+        """
+        # All custom serialization is done by the class provided to json.dump
+        json.dump(self._trained, stream, cls=TrainedSurrogateEncoder)
+
+    @classmethod
+    def load(cls, stream):
+        """Create an instance of a surrogate from a stream.
+
+            Args:
+                stream:
+                    This is the python stream containing the data required to load the surrogate.
+                    This is often, but does not need to be a string of json data.
+
+            Returns:
+                An instance of the derived class or None if it failed to load
+        """
+        stream.seek(0)
+        try:
+            json_obj = json.load(
+                stream, object_pairs_hook=TrainedSurrogateDecoder.decode_pairs
+            )
+            obj = PysmoSurrogate(
+                trained_surrogates=json_obj[TSEBase.MODEL_KEY],
+                input_labels=json_obj[TSEBase.INPUT_KEY],
+                output_labels=json_obj[TSEBase.OUTPUT_KEY],
+                input_bounds=json_obj[TSEBase.BOUNDS_KEY],
+            )
+        except JSONDecodeError as err:
+            obj = None
+            _log.error(f"Error decoding surrogate model (stream={stream}): {err}")
+        return obj
+
+
+# Serialization and deserialization classes
+# ------------------------------------------
+
+
+class TSEBase:
+    # top-level keys shared by encoder/decoder classes
+    MODEL_KEY = "model_encoding"
+    INPUT_KEY = "input_labels"
+    OUTPUT_KEY = "output_labels"
+    BOUNDS_KEY = "input_bounds"
+    TYPE_KEY = "surrogate_type"
+    MODEL_ATTR_KEY = "attr"
+    MODEL_MAP_KEY = "map"
+    MODEL_METRICS_KEY = "errors"
+
+
+class TrainedSurrogateEncoder(JSONEncoder, TSEBase):
+    # Attributes to encode
+    attrs = {
+        "final_polynomial_order",
+        "multinomials",
+        "optimal_weights_array",
+        "extra_terms_feature_vector",
+        "additional_term_expressions",
+        "regression_data_columns",
+        "errors",
+        "centres",
+        "x_data_columns",
+        "x_data_min",
+        "x_data_max",
+        "basis_function",
+        "self.sigma",
+        "y_data_min",
+        "y_data_max",
+        "weights",
+        "sigma",
+        "regularization_parameter",
+        "R2",
+        "rmse",
+        "optimal_weights",
+        "optimal_p",
+        "optimal_mean",
+        "optimal_variance",
+        "regularization_parameter",
+        "optimal_covariance_matrix",
+        "covariance_matrix_inverse",
+        "optimal_y_mu",
+        "training_R2",
+        "training_rmse",
+        "x_data",
+        "x_data_scaled",
+    }
+
+    def default(self, obj):
```

Comment:
should have a docstring saying this function is conforming to the interface expected by the json package

Reply:
Done.